pkg/sources/git/git.go

```diff
@@ -1253,9 +1283,44 @@ func (s *Git) handleBinary(ctx context.Context, gitDir string, reporter sources.
 		return fmt.Errorf("error starting git cat-file: %w\n%s", err, stderr.Bytes())
 	}
 
-	defer func() { _ = cmd.Wait() }()
+	// Ensure all data from the reader (stdout) is consumed to prevent broken pipe errors.
+	// This operation discards any remaining data after HandleFile completion.
+	// If the reader is fully consumed, the copy is essentially a no-op.
+	// If an error occurs while discarding, it will be logged and combined with any existing error.
+	// The command's completion is then awaited and any execution errors are handled.
+	defer func() {
+		n, copyErr := io.Copy(io.Discard, stdout)
+		if copyErr != nil {
+			ctx.Logger().Error(
+				copyErr,
+				"Failed to discard remaining stdout data after HandleFile completion",
+			)
+		}
+
+		ctx.Logger().V(3).Info(
+			"HandleFile did not consume all stdout data; excess discarded",
+			"bytes_discarded", n)
+
+		err = combineErrors(err, fmt.Errorf("error discarding excess stdout data: %w", copyErr))
+
+		// Wait for the command to finish and handle any errors.
+		waitErr := cmd.Wait()
+		err = combineErrors(err, fmt.Errorf("command execution error: %w", waitErr))
+	}()
+
+	return handlers.HandleFile(catFileCtx, stdout, chunkSkel, reporter, handlers.WithSkipArchives(s.skipArchives))
```

Comment:
Does `HandleFile` call `rdr.Close()` (and in this case `rdr` is `stdout`)? If yes, is it an issue to read from `stdout` after `Close()` has been called on it (in the `defer` block above)?

Reply:
It does, but the `HandleFile` function does not directly close `stdout`. Instead, it wraps `stdout` in a `BufferedReadSeeker`, and when `HandleFile` calls `rdr.Close()`, it only closes the `BufferedReadSeeker`, not `stdout`. The `Close()` method of `BufferedReadSeeker` manages internal resources like returning buffers to a pool or closing temporary files without affecting the underlying `stdout`. [here](https://github.com/trufflesecurity/trufflehog/blob/main/pkg/iobuf/bufferedreaderseeker.go#L371-L381)

So `stdout` remains open until `cmd.Wait()` completes in `handleBinary`. The deferred function in `handleBinary` ensures proper cleanup by reading any remaining data from `stdout` and waiting for the command to finish, which effectively closes `stdout`.

I feel like this approach effectively separates ownership: `HandleFile` manages the custom reader, while `handleBinary` is responsible for handling and closing `stdout` via `cmd.Wait()`.

Reply:
Okay, so `HandleFile` takes in an `io.Reader`, not `io.ReadCloser`. Closing is up to the caller.

Yeah, this is a great approach!

Reply:
It used to be an `io.ReadCloser`, but I confused myself a few times by mistakenly closing the passed-in reader and then trying to close it again at the call site. That's when I decided to change it to an `io.Reader` and have the callers handle closing. I think that's maybe why you also thought `rdr.Close()` closed the underlying reader.